env0/resource_cost_credentials.go

```diff
@@ -0,0 +1,208 @@
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"client_id", "client_secret", "tenant_id", "subscription_id", "arn", "external_id"},
+				ExactlyOneOf:  []string{"arn", "client_id"},
+			},
+			"secret": {
+				Type:          schema.TypeString,
+				Description:   "the secret of this credentials",
+				Optional:      true,
+				Sensitive:     true,
+				ForceNew:      true,
+				RequiredWith:  []string{"table_id"},
+				ConflictsWith: []string{"client_id", "client_secret", "tenant_id", "subscription_id", "arn", "external_id"},
+			},
+		},
+	}
+}
+
+func resourceCostCredentialsCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+
+	apiClient := meta.(ApiClientInterface)
+	var apikey client.Credentials
+	var err error
+	payLoad, credType, err := setPayload(d)
+	if err != nil {
+		return diag.Errorf("ERROR: %v", err)
+	}
+
+	switch credType {
+	case string(client.AwsCostCredentialsType):
+		payLoadToSent := payLoad.(client.AwsCredentialsCreatePayload)
+		payLoadToSent.Name = d.Get("name").(string)
+		apikey, err = apiClient.AwsCredentialsCreate(payLoadToSent)
+	case string(client.AzureCostCredentialsType):
+		payLoadToSent := payLoad.(client.AzureCredentialsCreatePayload)
+		payLoadToSent.Name = d.Get("name").(string)
+		apikey, err = apiClient.AzureCredentialsCreate(payLoadToSent)
+	case string(client.GoogleCostCredentiassType):
+		payLoadToSent := payLoad.(client.GoogleCostCredentialsCreatePayload)
+		payLoadToSent.Name = d.Get("name").(string)
+		apikey, err = apiClient.GoogleCostCredentialsCreate(payLoadToSent)
+	}
+
+	if err != nil {
+		return diag.Errorf("cost credential fialed: %v", err)
+	}
+	d.SetId(apikey.Id)
+	return nil
+}
+
+func resourceCostCredentialsRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	id := d.Id()
+	_, err := apiClient.CloudCredentials(id)
+	if err != nil {
+		return diag.Errorf("could not get credentials: %v", err)
+	}
+	return nil
+
+}
+
+func resourceCostCredentialsDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	id := d.Id()
+	err := apiClient.CloudCredentialsDelete(id)
+	if err != nil {
+		return diag.Errorf("could not delete credentials: %v", err)
+	}
+	return nil
+
+}
+
+func getCredType(d *schema.ResourceData) (string, error) {
+	_, awsOk := d.GetOk("arn")
+	_, azureOk := d.GetOk("client_id")
+	_, googleOk := d.GetOk("table_id")
+	switch {
+	case awsOk:
+		return string(client.AwsCostCredentialsType), nil
+	case azureOk:
+		return string(client.AzureCostCredentialsType), nil
+	case googleOk:
+		return string(client.GoogleCostCredentiassType), nil
+	default:
+		return "error", errors.New("error in schema, no required value  defined")
+	}
+
+}
+
+func setPayload(d *schema.ResourceData) (interface{}, string, error) {
```

Comment:
IMO `getCredentailsPayload` is better name

Reply:
e2814aded3ee3fac6539413714d2b80bb490692e